cert/spec.go

```diff
@@ -438,56 +365,114 @@ func (spec *Spec) ResetBackoff() {
 	}
+	if existingCert.NotBefore.After(now) {
+		// someone needs a better clock.
+		return fmt.Errorf("certificate isn't yet valid: %s", existingCert.NotBefore)
+	}
+	return nil
+}
+
+// EnforcePKI processes a spec, updating content on disk, taking action as needed.
+// Returns (TTL for PKI, error).  If an error occurs, the ttl is at best
+// a hint to the invoker as to when the next refresh is required- that said
+// the invoker should back off and try a refresh.
+func (spec *Spec) EnforcePKI(enableActions bool) error {
 
-	lifespan := time.Duration(0)
-	if !spec.Ready() {
-		log.Debugf("manager: %s isn't ready", spec)
+	updateReason := ""
+	var currentCA *x509.Certificate
+	var err error
+	if spec.renewalForced {
+		updateReason = "key"
 	} else {
-		log.Debugf("manager: %s checking lifespan", spec)
-		lifespan = spec.Lifespan()
-	}
-	log.Debugf("manager: %s has lifespan %s", spec, lifespan)
-	if lifespan <= 0 {
-		err := spec.RenewPKI()
+		currentCA, err = spec.CA.getRemoteCert()
 		if err != nil {
-			log.Errorf("manager: failed to renew %s; requeuing cert", spec)
-			return 0, err
+			log.Errorf("spec %s: failed getting remote: %s", spec, err)
+			return err
 		}
 
-		log.Debug("taking action due to key refresh")
-		if enableActions {
-			err = spec.TakeAction("key")
-		} else {
-			log.Infof("skipping actions for %s due to calling mode", spec)
+		if spec.CA.File != nil {
+			existingCA, err := spec.CA.File.ReadCertificate()
+			if err != nil {
+				log.Infof("spec %s: ca on disk needs regeneration: %s", spec, err)
+				updateReason = "CA"
+			} else {
+				spec.updateCAExpiry(existingCA.NotAfter)
+				if !existingCA.Equal(currentCA) {
+					log.Debugf("spec %s: ca has changed", spec)
+					updateReason = "CA"
+				} else {
+					log.Debugf("spec %s: ca is the same", spec)
+				}
+			}
 		}
 
-		// Even though there was an error managing the service
-		// associated with the certificate, the certificate has been
-		// renewed.
-		if err != nil {
-			metrics.ActionFailure.WithLabelValues(spec.Path, "key").Inc()
-			log.Errorf("manager: %s", err)
+		if updateReason == "" {
+			err := spec.checkDiskCertKey(currentCA)
+			if err != nil {
+				log.Infof("spec %s: forcing refresh due to %s", spec, err)
+				updateReason = "key"
+			}
 		}
+	}
 
-		log.Info("manager: certificate successfully processed")
+	if updateReason == "" {
+		log.Debugf("spec %s: still up to date", spec)
+		return nil
+	}
+
+	err = spec.renewPKI(currentCA)
+	if err != nil {
+		log.Errorf("manager: failed to renew %s; requeuing cert", spec)
+		return err
+	}
+
+	if enableActions {
+		err = spec.TakeAction(updateReason)
+	} else {
+		log.Infof("skipping actions for %s due to calling mode", spec)
+	}
+
+	// Even though there was an error managing the service
+	// associated with the certificate, the certificate has been
+	// renewed.
+	if err != nil {
+		metrics.ActionFailure.WithLabelValues(spec.Path, "key").Inc()
```

Comment:
No action for this PR, but just making note this is a place where metrics won't reflect actual state if there's no action failure issues.

Reply:
(as you mentioned in your PR, this is future work -- I'm just making note of it here)

Reply:
Yep, an immediate followup on my end is to move the action metric manipulation direction into 'takeAction' rather than the caller; that'll resolve this, including tracking actions taken.